### ISSUE TITLE
Allow deployment of modules through SDK

### ISSUE THREAD

**Issue by atbe** [nftlabs/nftlabs-sdk-ts#11]

**Comment by jnsdls** [comment]
interface we want:
```ts
public async deployModule(ModuleType, MetadataObjectRequiredToDeployModuleType>: Promise<ModuleOfModuleType>
```

**Comment by jakeloo** [comment]
we should define what the interfaces looks like for each of the module. All of our module should have `name, image, description` in the metadata. then, we need to map out what it would look like for royalty % and splits.

supports royalty for NFT, Bundle, LazyNFT. make it supports on OpenSea (in the json metadata) and our contract (in the constructor)
supports royalty splits for ^.

**Comment by atbe** [comment]
I'm personally a bigger fan of an interface that doesn't have a generic `deployModule` function and instead has strongly defined functions like this:

```typescript
import { BigNumberish } from "ethers";
import { ModuleMetadata, ModuleType, NFTModule } from "..";

interface BaseMetadata {
  name: string;
  description: string;
}

interface NftModuleMetadata extends BaseMetadata {
  image: string;
  royaltyBasisPoints: BigNumberish;
}

interface SplitsModuleMetadata extends BaseMetadata {
  image: string;

  // we could collect these two properties into a single interface, but
  // I chose to separate them out for clarity
  recipients: string[];
  recipientSplits: BigNumberish[];
}

export interface IAppModule {
  getAllModuleMetadata(
    filterByModuleType?: ModuleType[],
  ): Promise<ModuleMetadata[]>;

  deployNftModule(metadata: NftModuleMetadata): Promise<NFTModule>;

  // using `any` because SplitsModule isn't on this branch yet
  deploySplitsModule(metadata: SplitsModuleMetadata): Promise<any>;
}
```

**Comment by atbe** [comment]
The pain of using a single function to deploy a large set of module types is strong:

1. Changes to this function effect all callers
2. Generic interfaces that can't communicate to the caller whats *specific* about the module they're deploying (like an NFT/Collection/Pack having royalty configuration, but a currency does not)
3. Return type of the function would have to be a union type

**Comment by jnsdls** [comment]
> The pain of using a single function to deploy a large set of module types is strong:
> 
> 1. Changes to this function effect all callers
> 2. Generic interfaces that can't communicate to the caller whats _specific_ about the module they're deploying (like an NFT/Collection/Pack having royalty configuration, but a currency does not)
> 3. Return type of the function would have to be a union type

you can achieve the same thing with a single function by making it generic (aka based on the module type passed it can infer the type of metadata)

with that said I really do not care that much about that part, if it's separate functions and you feel strongly that it's easier for the user to understand that way then go for it

**Comment by jnsdls** [comment]
overall this looks like what I was expecting, if it's 1 fn vs 7 fns I don't *really* care (aside from if we add new modules it's 1 less step to add them (just add the type, no new functions to add to the public interfce)

**Comment by jakeloo** [comment]
```
  recipients: string[];
  recipientSplits: BigNumberish[];
```

I think these should be 1 interface array, it's an odd design to have people trying to align both array.
Also it's not really clear what's supposed to go into Splits. I wonder if there's anything more explicit we could name it?

**Comment by jakeloo** [comment]
```
interface NftModuleMetadata extends BaseMetadata {
  image: string;
  royaltyBasisPoints: BigNumberish;
}
```

I think optional royalty recipient should probably be in here:
1) if it's null = the project address
2) royalty split address. 

the contract today does check for (2), one idea is that we could do:
`protocolControl.readOnlyContract.staticcall.setModuleRoyaltyTreasury(...)` -- since it's a static call, it won't actually execute, but it'd check if the app contract would take in the royalty treasury address or not.

In the contract layer, we basically check for 5% our fee.

**Comment by jakeloo** [comment]
Probably only NFT modules need `image` in the metadata.

**Comment by atbe** [comment]
> I think these should be 1 interface array, it's an odd design to have people trying to align both array.
Also it's not really clear what's supposed to go into Splits. I wonder if there's anything more explicit we could name it?

So that was just pseudocode for this conversation, the actual implementation uses a strict interface. You can find it [here](https://github.com/nftlabs/nftlabs-sdk-ts/pull/24/files#diff-a1cc0c406593f1a9c9ca37a5df56db5fbbaa9fbe9421035af48b245ffafc3ba4R16)

> I think optional royalty recipient should probably be in here:

I agree, and its already there, you can find the property [here](https://github.com/nftlabs/nftlabs-sdk-ts/pull/24/files#diff-06b285065734cb047634f42735aa6a363ea1382ae7958f7951e5d4a28b44b6deR30).

I don't like option 2 because its implicit behavior that we would have to communicate in a comment or in the documentation. On that note though, it would be rare (and maybe even erroneous) for someone to *not* pass in a fee recipient address because they could have forgotten to set it and then they lose royalties (is that what happens when there's a resale and the recipient address is undefined?)

> Probably only NFT modules (NFT, Bundle, Pack, Drops) need image in the metadata.

Yeah that's what I thought as well. The PR is where you should do reviews now, so check that one out. I completed the implementations for the Collection and Splits deployment. Will complete the rest of the modules today.

**Comment by atbe** [comment]
This is complete and has been released in version `v1.17.0`. Excited to hear some feedback from users!

**Comment by jnsdls** [comment]
@atbe will integrate into console today
